### PR TITLE
Remove Exam

### PR DESCRIPTION
**Pull request by lotrfan**
Potential areas regarding more discussion:
- `RetryExam`s don't have a `nQuestions` option - when converting, I pretty much just ignore it (I do issue a warning if the `Exam`'s `nQuestions` doesn't match the number of questions in the `RetryExam`)
- https://github.com/lotrfan/PrairieLearn/blob/3a03bb4ca1d1dfde768dbf6fbb1f6316e0701ef5/backend/RetryExamTestServer.js#L199 - if we do want a call to `checkObjAuth`, we'll need to pass along the request to `updateTInstance`

Testing:
- If you don't have any in-progress or completed `Exam`-type `tInstance`s in the database, you'll have to do those off of `remove-exam~~` (or master...)
- Here are the two `info.json`'s I used: https://gist.github.com/lotrfan/839f7793b412d8617d15

Also, the first `info.json` in the gist, with `nQuestions: 5`, doesn't always create an exam with 5 questions---sometimes, there were only 4. The maximum grade of such instances was 80%, since the test's maximum score, which is what was used to calculate the percentage, was 5. Should we be using `tInstance.maxScore` instead of `test.maxScore` in the other test types?

### CODE REVIEW

**Comment by mwest1066** [comment]
+117 lines, -542 lines: this is the best kind of patch!

**Comment by mwest1066** [comment]
RetryExams actually do have an nQuestions option. This is all a bit horrible, but RetryExams allow two different ways of specifying the question pool from which they randomize.

The "new" way of specifying the question pool is with zones, each of which has a list of questions, each of which has a list of QIDs, and one of these QIDs is selected per question. This doesn't need an `nQuestions` parameter because we always have all the questions, and it's just the QID choice per question that is random.

The "old" way of specifying the question pool is to have an array of arrays of arrays of `(qid, points)` pairs, together with an overall `nQuestions`. Exactly `nQuestions` should be chosen from the big list, attempting to balance the number chosen from each array entry at each nesting level. The idea behind the 3-fold nested arrays is that this allows QIDs to be grouped by difficulty and topic, and then QID selection is balanced between difficulties and between topics.

I apologize for the fact that only the "new" format is [described in the docs](https://github.com/PrairieLearn/PrairieLearn/blob/master/doc/writingTests.md#retryexam-randomization). I had been intending to remove the "old" way, but it's helpful for making practice exams where you just take all the questions from all homeworks, roughly sort them by difficulty, and then allow the randomization to generate practice exams from a huge question pool.

This "old" style is also helpful right now because we can auto-convert `Exam` to `RetryExam` with no loss of information. I think the only change is that `Exam` uses `qidGroups` as nested arrays of QIDs, while `RetryExam` has `questionGroups` as nested arrays of `(QID, points)` pairs. `Exam` simply assumed that all questions had 1 point.

Both "new" and "old" styles of RetryExams are specified in [testOptionsRetryExam.json`](https://github.com/PrairieLearn/PrairieLearn/blob/master/backend/schemas/testOptionsRetryExam.json), but we use [`oneOf`](http://json-schema.org/latest/json-schema-validation.html#anchor88) to ensure that we only have [either "new" or "old" specifications](https://github.com/PrairieLearn/PrairieLearn/blob/master/backend/schemas/testOptionsRetryExam.json#L63).

**Comment by mwest1066** [comment]
I think it's fine to skip the call to `checkObjAuth()`. All calls to `updateTInstance()` already checked that we have auth for the `tInstance` or `test`, so there shouldn't be any obvious hole here. Normally I would advocate checking auth again anyway, just to provide defense in depth, but given that this is temporary upgrade code I think it's not worth adding complexity.

**Comment by mwest1066** [comment]
The fact that [`info.exam.json`](https://gist.github.com/lotrfan/839f7793b412d8617d15) sometimes generates tests with 4 questions is a bug. I added #308 and I'll try and fix it now.

**Comment by lotrfan** [comment]
Is the old way going to go away? [The comments](https://github.com/lotrfan/PrairieLearn/blob/3a03bb4ca1d1dfde768dbf6fbb1f6316e0701ef5/backend/RetryExamTestServer.js#L123) seem to indicate that zones are the way forward.

**Comment by mwest1066** [comment]
I think we should get rid of the old way at some point, but for now I think we should auto-convert `Exam` to the old type of `RetryExam` because it's a lossless conversion. We can continue to support this old type during this semester, but then remove it after semester ends. I'm generally trying to avoid making breaking changes during semester (deprecating things is fine, just not making changes that actually require instructors to fix things).

**Comment by mwest1066** [comment]
After this PR is merged we'll have just a single exam-style test type. Should we call this `RetryExam` or `Exam`? At some point I think it should be called just plain `Exam`, but I'm not sure whether it's less confusion to do that immediately or to wait and do it with another deprecation warning at some point in the future.

**Comment by lotrfan** [comment]
Are the `points` per question variant in a `questionGroups`? I.e., would this `Exam`:

``` json
"qidGroups": [
    [
        ["fossilFuelsRadio"],
        ["particleMotion"],
        ["fossilFuelsTrueFalse"]
    ], [
        ["fibonacci", "addVectors"],
        ["fossilFuelsCheckbox"]
    ]
]
```

correspond to this `RetryExam`?

``` json
"questionGroups": [
    [
        [ {"qid": "fossilFuelsRadio", "points": [1]} ],
        [ {"qid": "particleMotion", "points": [1]} ],
        [ {"qid": "fossilFuelsTrueFalse", "points": [1]} ]
    ],
    [
        [
            {"qid": "fibonacci", "points": [1]},
            {"qid": "addVectors", "points": [1]}
        ],
        [ {"qid": "fossilFuelsCheckbox", "points": [1]} ]
    ]
]
```

(Particularly with respect to the question with `fibonacci` and `addVectors`.)

**Comment by mwest1066** [comment]
That looks correct to me. It's also correct according to [the schema](https://github.com/PrairieLearn/PrairieLearn/blob/master/backend/schemas/testOptionsRetryExam.json#L17).

**Comment by lotrfan** [comment]
cool... more commits on their way :)

**Comment by lotrfan** [comment]
If those look OK, I can squash them before merging

File: backend/RetryExamTestServer.js [hunk 1]
```diff
@@ -179,7 +181,61 @@ define(["underscore", "moment-timezone", "PrairieRandom"], function(_, moment, P
                 qid: qid,
             });
         });
-        return tInstance;
+
+        // convert old Exam tInstances to RetryExams
+        if (tInstance.questionsByQID === undefined) {
+            tInstance.zones = [null]; // put everything into one zone
```
Comment (mwest1066) [inline backend/RetryExamTestServer.js]: By comparison with the code around L123, I think we just don't need to set `zones` or related properties for "old-style" `RetryExam`s.

File: backend/RetryExamTestServer.js [hunk 2]
```diff
@@ -179,7 +181,61 @@ define(["underscore", "moment-timezone", "PrairieRandom"], function(_, moment, P
...
+            if (!tInstance.submissionsByQid) tInstance.submissionsByQid = {};
+
+            async.each(tInstance.qids, function(qid, cb) {
+                if (!tInstance.qiidsByQid) return cb("No qiidsByQid");
+                var qiid = tInstance.qiidsByQid[qid];
+                if (!qiid) return cb("No qiidsByQid.qid for qid: " + qid);
+                db.qiCollect.findOne({qiid: qiid}, function(err, qInstance) {
+                    if (err) return cb(err);
+                    if (!qInstance) return cb("No qInstance with qiid " + qiid);
+                    // TODO: should there be a checkObjAuth here?
```
Comment (mwest1066) [inline backend/RetryExamTestServer.js]: https://github.com/PrairieLearn/PrairieLearn/pull/306#issuecomment-192690597

**Comment by mwest1066** [comment]
Looks great. I'm just doing actual checks on running code now.

**Comment by mwest1066** [comment]
I tested on CS 225 `hourExam1` and it upgraded perfectly.

Do we want to try and upgrade tInstances proactively on server load? Is there some reasonably fast way to determine if this is necessary? If we did that then we would be able to also remove code like https://github.com/PrairieLearn/PrairieLearn/blob/master/backend/server.js#L2373

**Comment by lotrfan** [comment]
The "nicest" way would probably be something like (pseudocode):

``` js
db.tCollect.find({"type": "Exam"}, function(test) {
  db.tiCollect.find({"tid": test.tid}, function(tInstance) {
    updateTInstance(test, tInstance);
  });
});
```

Not sure sure how fast it would be though...

**Comment by mwest1066** [comment]
I think upgrade on server start would be good, actually, because then we can just run this once and then delete all this tinstance upgrade code. Even if it's slow, we only need to run it briefly.

**Comment by lotrfan** [comment]
alright... don't have time to work on that now, but maybe later tonight or tomorrow.

Do you think it's worth:
- saving a backup of the old `tInstance` (in, say, a field of the updated `tInstance`)?
- having the `test.type` stored in the `tInstance`?

**Comment by mwest1066** [comment]
I think saving a copy of the old `tInstance` is a great idea just in case we need it for some reason later. There is a special collection (`dCollect`) for deleted objects, and a function `deleteObjects()` that might be useful here. Storing it directly in the updated `tInstance` is also fine.

If it would make things easier to store `test.type` in the `tInstance` then I'm fine with it. I guess in general I feel that the current DB schema (such as it is) is horribly non-normalized, with duplicate information all over the place. Perhaps surprisingly this hasn't actually caused many problems so far, but my painful experience from other projects suggests that replicating data is a source of pain sooner or later.

**Comment by mwest1066** [comment]
Closing without merging as this has been superseded by v2, which implemented this in 4544ac3e4be4119294522005ca3e77488bdda233

**Pull request closed without merging**